ada_feeding_perception/ada_feeding_perception/segment_all_items.py

```diff
@@ -0,0 +1,1279 @@
+        )
+
+        # Publish a visualization of the GroundingDINO predictions, if the visualization
+        # flag is set to true
+        if self.viz_groundingdino:
+            self.visualize_groundingdino_results(image, bbox_predictions)
+
+        # Collect the top contender mask for each food item label detected by
+        # GroundingDINO using EfficientSAM and create dictionary of mask
+        # predictions from the pipeline
+        detected_items = []
+        item_labels = []
+        masks_list = []
+        mask_num = 1
+        for phrase, boxes in bbox_predictions.items():
+            for box in boxes:
+                masks, scores = self.run_efficient_sam(image, None, box, 1)
+                if len(masks) > 0:
+                    masks_list.append(masks[0])
+                    item_id = f"food_id_{mask_num:d}"
+                    mask_num += 1
+                    mask_msg = self.generate_mask_msg(
+                        item_id, phrase, scores[0], masks[0], image, depth_img, box
+                    )
+                    detected_items.append(mask_msg)
+                    item_labels.append(phrase)
+
+        result.detected_items = detected_items
+        result.item_labels = item_labels
+
+        # Measure the elapsed time running GroundingDINO on the image prompt
+        inference_time = int(round((time.time() - inference_time) * 1000))
+        self._node.get_logger().info(
+            f"Approximate Vision Pipeline Inference Time: {inference_time}"
+        )
+
+        return result
+
+    async def execute_callback(
+        self, goal_handle: ServerGoalHandle
+    ) -> SegmentAllItems.Result:
+        """
+        Execute the action server callback.
+
+        Parameters
+        ----------
+        goal_handle: The goal handle for the action server.
+
+        Returns
+        -------
+        result: The result message containing masks for all food items detected in the image
+                paired with semantic labels.
+        """
+        self._node.get_logger().info("Received a new goal!")
+        starting_time = self._node.get_clock().now()
+
+        # Get the latest image and camera info
+        latest_img_msg = self._node.get_latest_msg(self.rgb_image_topic)
+        if self.camera_info is None:
+            self.camera_info = self._node.get_latest_msg(self.camera_info_topic)
+        if self.camera_info is not None:
+            camera_info = self.camera_info
+        else:
+            camera_info = None
+
+        # Check if the image and camera info are available
+        if latest_img_msg is None or camera_info is None:
+            self._node.get_logger().error("Image or camera info not available.")
+            return SegmentAllItems.Result()
+
+        # Get the caption from the goal request
+        caption = goal_handle.request.caption
+
+        # Create a rate object to control the rate of the vision pipeline
+        rate = self._node.create_rate(self.rate_hz)
```

Comment:
I recently learned that rates live forever in a node unless explicitly deleted by `destroy_rate` (see comment [here](https://github.com/ros2/rclpy/blob/78f5e14cdabeb047ab39829caef7a4872f3179ff/rclpy/rclpy/node.py#L1891)). So I'd recommend adding that at the end of this function, else the rate will keep taking up callback resources even after the action has finished.